### DISCUSSION TITLE
Plugin-typing and Grouping AaC Commands

### DISCUSSION THREAD

**Discussion by Coffee2Bits** [DevOps-MBSE/AaC#224] [Ideas]
As we're building more and more functionality for AaC, we're starting to see a negative to our plugin-centric architecture which is an ever-increasing list of available commands for the AaC package.

For example, earlier versions such as [6ed806531671cb6129efab59f164043336a0840f](https://github.com/jondavid-black/AaC/tree/6ed806531671cb6129efab59f164043336a0840f) (Nov 15) 
```bash
(venv) gitpod /workspace/AaC ((6ed8065...)) $ aac -h
usage: aac [-h] {validate,aac-core-spec,gen-plugin,json} ...

positional arguments:
  {validate,aac-core-spec,gen-plugin,json}
    validate            Ensures the AaC yaml is valid per the AaC core spec
    aac-core-spec       Prints the AaC model describing core AaC data types and enumerations
    gen-plugin          Generates an AaC plugin from an AaC model of the plugin
    json                Converts an AaC model to JSON

optional arguments:
  -h, --help            show this help message and exit
```

Whereas master gives the user this:
```bash
(venv) gitpod /workspace/AaC/python (main) $ aac -h
usage: aac [-h] {validate,version,aac-core-spec,spec-validate,puml-component,puml-sequence,puml-object,gen-gherkin-behaviors,gen-design-doc,gen-protobuf,gen-plugin,gen-json} ...

positional arguments:
  {validate,version,aac-core-spec,spec-validate,puml-component,puml-sequence,puml-object,gen-gherkin-behaviors,gen-design-doc,gen-protobuf,gen-plugin,gen-json}
    validate                  Ensures the AaC yaml is valid per the AaC core spec
    version                   Outputs Architecture-as-Code Python package version
    aac-core-spec             Prints the AaC model describing core AaC data types and enumerations
    spec-validate             Validates spec traces within the AaC model.
    puml-component            Converts an AaC model to a Plant UML component diagram.
    puml-sequence             Converts an AaC usecase to a Plant UML sequence diagram.
    puml-object               Converts an AaC model to a Plant UML object diagram.
    gen-gherkin-behaviors     Generate Gherkin feature files from Arch-as-Code model behavior scenarios.
    gen-design-doc            Generate a System Design Document from Architecture-as-Code models.
    gen-protobuf              Generate protobuf messages from Arch-as-Code models.
    gen-plugin                Generates an AaC plugin from an AaC model of the plugin
    gen-json                  Converts an AaC model to JSON

optional arguments:
  -h, --help                  show this help message and exit
```

So, given a few more months of plugin development, we can add a lot more to this, but at what point do we start to overwhelm users? 

I can think of a few strategies to address this:
1. Allow users to activate and deactivate plugins (only activate plugins show up in the help command)
2. We group plugins by common functionality (Diagramming, Generation, Documentation, etc..)

### Activating and Deactivating Plugins
If we allow users to activate and deactivate plugins, then we'll need to store the active state of plugins -- this is something we can leverage those workspace configurations for. Deactivated plugins can be listed via a command like `aac plugin show-inactive` which allows users to see all available options and curate the plugins they interact with.

### Grouping Plugin Types
If we group the commands by common type like diagramming, documentation, infrastructure, service-stubbing, and whatever other categories we come up with, then we can leverage `argparse`'s [group functionality](https://docs.python.org/3/library/argparse.html#argument-groups). Argparse's group functionality would let us produce something like:

```bash
(venv) gitpod /workspace/AaC/python (main) $ aac -h
usage: aac [-h] {validate,version,aac-core-spec,spec-validate,puml-component,puml-sequence,puml-object,gen-gherkin-behaviors,gen-design-doc,gen-protobuf,gen-plugin,gen-json} ...

positional arguments:
  {validate,version,aac-core-spec,spec-validate,puml-component,puml-sequence,puml-object,gen-gherkin-behaviors,gen-design-doc,gen-protobuf,gen-plugin,gen-json}

optional arguments:
  -h, --help                  show this help message and exit

Core:
  These commands are core to AaC.

    validate                  Ensures the AaC yaml is valid per the AaC core spec
    version                   Outputs Architecture-as-Code Python package version
    aac-core-spec             Prints the AaC model describing core AaC data types and enumerations

Diagramming:
  These plugins provide the ability to generate diagrams from AaC models.

    puml-component            Converts an AaC model to a Plant UML component diagram.
    puml-sequence             Converts an AaC usecase to a Plant UML sequence diagram.
    puml-object               Converts an AaC model to a Plant UML object diagram.

Validation:
  These plugins provide the ability to validate and verify the content of models.

  spec-validate             Validates spec traces within the AaC model.

Generation:
  These plugins provide the ability to generate artifacts from AaC models and definitions

    gen-gherkin-behaviors     Generate Gherkin feature files from Arch-as-Code model behavior scenarios.
    gen-design-doc            Generate a System Design Document from Architecture-as-Code models.
    gen-protobuf              Generate protobuf messages from Arch-as-Code models.
    gen-plugin                Generates an AaC plugin from an AaC model of the plugin
    gen-json                  Converts an AaC model to JSON

```

### Alex's Personal Preference
I personally like both of these options and their application to more than just this concern of organizing help command output. Activating and deactivating plugins will help us cut down the noise AaC users will experience while still allowing them to configure their personal and workspace configurations. Grouping commands will allow us to investigate reducing plugin boilerplate and providing plugin classes (GenerationPlugin, DocumentationPlugin, ValidationPlugin, etc).

**Comment by jondavid-black** [comment]
I like the idea of argument groups and the cleaner version of help it provides.  It seems like if we do this, we could still provide a workspace control later to enable/disable commands if we find it to be valuable.

**Reply by Coffee2Bits** [reply]
Definitely. If we allow users to enable/disable plugins, then that's something we will want to expose in the workspace configuration, especially since deactivating a plugin can break a model that's considered valid.

**Comment by cvchaparro** [comment]
I also like the idea of grouping commands.

Question on the workspace configuration - is this specific to a VSCode workflow? If so, are we going to have an alternative for users that don't want to use VSCode for model development (i.e. because they have another tool they prefer using for creating models, etc), or are we taking the stance that we're only supporting a VSCode workflow (for now?) when creating models, etc?

Finally, if we support more than just VSCode (or even if we don't support it but a user wants to use something other than VSCode) will there be a way for them to apply those same configurations without having to depend on VSCode, or would they effectively be out of luck on that front?

**Reply by jondavid-black** [reply]
I don't think it should be specific to a VSCode usage scenario.  I was thinking of it more like git config.

**Comment by Coffee2Bits** [comment]
With the conversations about leveraging a pattern akin to the contribution points in VSCode extensions, the more appropriate path forward might be allowing plugins to define tags for commands.

For instance, a generation command could use the tag 'generation' to group itself with other generation commands. We can provided tags with a little more specificity such as 'diagram', 'code-generation', 'linting', etc. 

This wouldn't group plugins so much as their commands, since plugins will have so much flexibility with the contribution points/various plugin hooks that it would be difficult to group plugins based on the whole breadth of their functionality.

**Reply by jondavid-black** [reply]
Do we need to be more explicit with the definition of plugins vs models?  Should there be a `plugin` root type with plugin specific field definitions?  How would someone create a plugin and define the group the plugin or even specific plugin commands belong to?

**Reply by Coffee2Bits** [reply]
I like the idea of a `plugin` root key; I think it would be an important distinction from models.

We could leverage tags to group commands. We can provide the list of available tags via an enum, like we do with Primitive field types.

**Reply by cvchaparro** [reply]
Is there a reason to make it a root key as opposed to making `plugin` an extension of `model`? Couldn't we achieve the desired functionality (i.e. grouping, etc) by having those fields on the extension?
